C/plugins/storage/postgres/connection.cpp

```diff
@@ -2546,3 +2546,87 @@ void Connection::logSQL(const char *tag, const char *stmt)
 		Logger::getLogger()->info("%s: %s", tag, stmt);
 	}
 }
+
+/**
+ * Create snapshot of a common table
+ *
+ * @param table		The table to snapshot
+ * @param id		The snapshot id
+ * @return		-1 on error, >= 0 on success
+ *
+ * The new created table name has the name:
+ * table_id
+ */
+int Connection::create_table_snapshot(const string& table, const string& id)
+{
+	string query = "SELECT * INTO TABLE foglamp.";
+	query += table + "_" +  id + " FROM foglamp." + table;
+
+	logSQL("CreateTableSnapshot", query.c_str());
+
+	PGresult *res = PQexec(dbConnection, query.c_str());
+	if (PQresultStatus(res) == PGRES_COMMAND_OK)
+	{
+		PQclear(res);
+		return 1;
+	}
+
+	raiseError("create_table_snapshot", PQerrorMessage(dbConnection));
+	PQclear(res);
+	return -1;
+}
+
+/**
+ * Set the contents of a common table from a snapshot
+ *
+ * @param table		The table to fill
+ * @param id		The snapshot id of the table
+ * @return		-1 on error, >= 0 on success
+ *
+ */
+int Connection::load_table_snapshot(const string& table, const string& id)
+{
+	string query = "INSERT INTO foglamp." + table;
+	query += " SELECT * FROM foglamp." + table + "_" + id;
```

Comment:
Though not sure, but it seems data from snapshot table is being appended to original table? Is it possible that there might be duplicate entries after the append operation?

Reply:
The command above creates a new table "aaa_123" with the content of "aaa"

Reply:
Isn't this restore function (load_table_snapshot) - bringing data back from snapshot table to original table? I am saying original table might have some data already. Is that possible?

Reply:
Oh yes.

If you try to "restore" configuration table, currently not empty, there will be some insert failures

Reply:
The operation PUT then returns the error message from database.
Same for POST (existing snapshot) or DELETE (non existing snapshot): errors are rerturned